cards/git-realm/writer.ts

```diff
@@ -0,0 +1,401 @@
+import { extractSettings } from './lib/git-settings';
+// import { PrimaryData } from 'jsonapi-typescript';
+
+import { Writer } from '@cardstack/core/writer';
+import { Session } from '@cardstack/core/session';
+import { UpstreamDocument, UpstreamIdentity, upstreamIdToCardId, upstreamIdToString } from '@cardstack/core/document';
+import { inject } from '@cardstack/hub/dependency-injection';
+import { AddressableCard } from '@cardstack/core/card';
+
+// import { Session } from '@cardstack/core/session';
+// import { UpstreamDocument, UpstreamIdentity } from '@cardstack/core/document';
+// import { inject } from '@cardstack/hub/dependency-injection';
+// import { AddressableCard } from '@cardstack/core/card';
+// import CardstackError from '@cardstack/core/error';
+
+// let counter = 0;
+
+// export default class EphemeralWriter implements Writer {
+//   ephemeralStorage = inject('ephemeralStorage');
+
+//   constructor(private realmCard: AddressableCard) {}
+
+//   async create(_session: Session, doc: UpstreamDocument, upstreamId: UpstreamIdentity | null) {
+//     let id = upstreamId ?? String(counter++);
+//     let saved = this.ephemeralStorage.store(doc, id, this.realmCard.csId);
+//     return { saved: saved!, id };
+//   }
+
+//   async update(_session: Session, id: UpstreamIdentity, doc: UpstreamDocument) {
+//     let version = doc.jsonapi.data.meta?.version;
+//     if (version == null) {
+//       throw new CardstackError('missing required field "meta.version"', {
+//         status: 400,
+//         source: { pointer: '/data/meta/version' },
+//       });
+//     }
+
+//     return this.ephemeralStorage.store(doc, id, this.realmCard.csId, String(version))!;
+//   }
+
+//   async delete(_session: Session, id: UpstreamIdentity, version: string | number) {
+//     this.ephemeralStorage.store(null, id, this.realmCard.csId, version);
+//   }
+// }
+
+import stringify from 'fast-json-stable-stringify';
+
+const defaultBranch = 'master';
+
+function getType(document: UpstreamDocument) {
+  return document.jsonapi.data.type;
+}
+
+function getId(document: UpstreamDocument) {
+  return document.jsonapi.data.id;
+}
+
+// function getMeta(model: todo) {
+//   return model.data ? model.data.meta : model.meta;
+// }
+
+// interface WriterConfig {
+//   repo: string;
+//   basePath?: string;
+//   branchPrefix?: string;
+//   remote?: RemoteConfig;
+// }
+
+export default class GitWriter implements Writer {
+  cards = inject('cards');
+
+  myEmail: string;
+  myName: string;
+  remote?: RemoteConfig;
+  repoPath: string;
+  basePath?: string;
+  repo?: Repository;
+  branchPrefix: string;
+  githereumConfig: todo;
+  githereum: todo;
+  _githereumPromise?: Promise<todo>;
+
+  constructor(private realmCard: AddressableCard) {
+    let hostname = os.hostname();
+    this.myName = `PID${process.pid} on ${hostname}`;
+    this.myEmail = `${os.userInfo().username}@${hostname}`;
+
+    let settings = extractSettings(realmCard);
```

Comment:
this will be async after you start using the public API on `AddressableCard` which means you wont be able to set this in the contructor. 

Our container actually uses async instantiation, so we do have an async intantiation hook for you to use for this kind of stuff: `async ready(){ ...} `

as part of the way the container will instantiate these features, it will await the `ready()` method in your writer immedately after instantiating the class (and providing injections as well into your instantiation). you put any async that you need as part of your writer there. So you'll need to have something like:
```ts
async ready(): Promise<void> {
  let settings = await extractSetting(realmCard);
  this.repoPath = settings.repo;
  this.basePath = settings.basePath;
  this.branchPrefix = settings.branchPrefix;
  this.remote = settings.remote;
}
```